python_modules/dagster/dagster/_core/utils.py

```diff
@@ -127,6 +128,35 @@ def parse_env_var(env_var_str: str) -> Tuple[str, str]:
 class InheritContextThreadPoolExecutor(ThreadPoolExecutor):
     """A ThreadPoolExecutor that copies over contextvars at submit time."""
 
+    def __init__(
+        self,
+        max_workers: int | None = None,
+        thread_name_prefix: str = "",
+        initializer: Callable[..., object] | None = None,
+        initargs: tuple[Any, ...] = ...,
+    ) -> None:
+        super().__init__(max_workers, thread_name_prefix, initializer, initargs)
+        self._futures = []
+
     def submit(self, fn, *args, **kwargs):
         ctx = copy_context()
-        return super().submit(ctx.run, fn, *args, **kwargs)
+        future = super().submit(ctx.run, fn, *args, **kwargs)
+        self._futures = [
+            future for future in self._futures if not future.done()
+        ]  # clean up done futures
```

Comment:
nit: `future` symbol used twice (new and in list comprehension), do `new_future` or something